File: backend/fly_pilot/protocol.py
"""JSON protocol between the browser and the Python sandbox."""

from __future__ import annotations

import json
from typing import Any

from fly_pilot.runway import Runway
from fly_pilot.sandbox import LandingSandbox, SandboxSnapshot
from fly_pilot.state import AircraftControls


def parse_client_message(raw: str | bytes) -> dict[str, Any]:
    if isinstance(raw, bytes):
        raw = raw.decode("utf-8")
    data = json.loads(raw)
    if not isinstance(data, dict) or "type" not in data:
        raise ValueError("client message must be an object with a type field")
    return data


def controls_from_message(data: dict[str, Any]) -> AircraftControls:
    return AircraftControls(
        aileron=float(data.get("aileron", 0.0)),
        elevator=float(data.get("elevator", 0.0)),
        rudder=float(data.get("rudder", 0.0)),
        throttle=float(data.get("throttle", 0.4)),
    ).clamped()


def error_payload(code: str, message: str, action: str | None = None) -> dict[str, Any]:
    return {
        "type": "error",
        "code": code,
        "message": message,
        "action": action,
    }


def hello_payload(sandbox: LandingSandbox) -> dict[str, Any]:
    runway: Runway = sandbox.runway
    approach = sandbox.approach
    mode = getattr(sandbox, "mode", sandbox.controller.name)
    observing = bool(getattr(sandbox, "observing", False))
    if mode == "fly_control":
        note = (
            "FLY CONTROL: fixed MaleCNS LIF plus a trained temporal decoder writes "
            "JSBSim inceptors. This is not biological synaptic learning. "
            "ExpertLandingController is not in the loop. Decoder input is DN activity only."
        )
    elif mode == "expert_observing":
        note = (
            "EXPERT + FLY OBSERVING: ExpertLandingController (conventional autopilot) "
            "flies the Cessna. MaleCNS watches a rendered fly-view. The fly is NOT "
            "controlling the aircraft."
        )
    elif sandbox.controller.name == "expert":
        note = (
            "ExpertLandingController is a conventional classical autopilot used as a "
            "solvability baseline and expert-data generator. It is not MaleCNS and is "
            "not biological computation."
        )
    else:
        note = (
            "Milestone 5 landing sandbox. MANUAL and EXPERT are unchanged. "
            "FLY CONTROL uses frozen MaleCNS plus an external trained decoder."
        )
    fly_controls = mode == "fly_control"
    payload: dict[str, Any] = {
        "type": "hello",
        "milestone": 5,
        "controller": mode,
        "control_authority": sandbox.controller.name,
        "physics": "jsbsim",
        "aircraft": "c172p",
        "runway": {
            "name": runway.name,
            "lat_deg": runway.lat_deg,
            "lon_deg": runway.lon_deg,
            "alt_m": runway.alt_m,
            "heading_deg": runway.heading_deg,
            "length_m": runway.length_m,
            "width_m": runway.width_m,
        },
        "approach": {
            "distance_m": approach.distance_m,
            "agl_m": approach.agl_m,
            "airspeed_kts": approach.airspeed_kts,
            "flight_path_deg": approach.flight_path_deg,
        },
        "integrity": {
            "authoritative_physics": "JSBSim Cessna 172P",
            "visuals": "Three.js human view; Python cubemap is canonical MaleCNS input",
            "male_cns": fly_controls or observing,
            "male_cns_observing": observing,
            "fly_controls_aircraft": fly_controls,
            "expert_is_biological": False,
            "biological_learning": False,
            "decoder_input": "dn_windowed_rates" if fly_controls else None,
            "note": note,
        },
        "schedule": {
            "physics_hz": 120.0,
            "vision_hz": 50.0,
            "neural_hz": 50.0,
            "state_broadcast_hz": 30.0,
        },
        "capabilities": (
            sandbox.mode_capabilities()
            if callable(getattr(sandbox, "mode_capabilities", None))
            else {}
        ),
    }
    observer = getattr(sandbox, "observer", None)
    if observer is not None and (observing or fly_controls):
        payload["observing"] = observer.metadata()
        payload["observing"]["controls_aircraft"] = fly_controls
    return payload


def state_payload(snapshot: SandboxSnapshot) -> dict[str, Any]:
    obs = snapshot.observation
    ep = snapshot.episode
    extra = obs.extra or {}
    mode = snapshot.ui_mode
    payload: dict[str, Any] = {
        "type": "state",
        "sim_time": obs.sim_time_s,
        "paused": snapshot.paused,
        "controller": mode,
        "control_authority": snapshot.controller_name,
        "observing": snapshot.observing,
        "spawn_seed": snapshot.spawn_seed,
        "position": {
            "lat_deg": obs.lat_deg,
            "lon_deg": obs.lon_deg,
            "alt_msl_m": obs.alt_msl_m,
            "alt_agl_m": obs.alt_agl_m,
            "east_m": obs.east_m,
            "north_m": obs.north_m,
            "up_m": obs.up_m,
            "along_m": obs.along_m,
            "right_m": obs.right_m,
        },
        "velocity": {
            "airspeed_kts": obs.airspeed_kts,
            "groundspeed_kts": obs.groundspeed_kts,
            "vertical_speed_fpm": obs.vertical_speed_fpm,
        },
        "attitude": {
            "pitch_deg": obs.pitch_deg,
            "roll_deg": obs.roll_deg,
            "heading_deg": obs.heading_deg,
            "alpha_deg": obs.alpha_deg,
            "p_deg_s": obs.p_deg_s,
            "q_deg_s": obs.q_deg_s,
            "r_deg_s": obs.r_deg_s,
            "beta_deg": obs.beta_deg,
        },
        "controls": {
            "aileron": obs.aileron,
            "elevator": obs.elevator,
            "rudder": obs.rudder,
            "throttle": obs.throttle,
        },
        "surfaces": {
            "elevator_pos": extra.get("elevator_pos", obs.elevator),
            "aileron_pos": extra.get("aileron_pos", obs.aileron),
            "rudder_pos": extra.get("rudder_pos", obs.rudder),
            "throttle_pos": extra.get("throttle_pos", obs.throttle),
        },
        "gear": {
            "on_ground": obs.on_ground,
            "wow": list(obs.wow),
        },
        "episode": {
            "status": ep.status.value,
            "reason": ep.reason,
            "touchdown_fpm": ep.touchdown_fpm,
        },
    }
    if snapshot.controller_name == "expert" and snapshot.debug:
        payload["expert"] = snapshot.debug
    if snapshot.ui_mode == "fly_control":
        payload["fly_control"] = {
            k: snapshot.debug[k]
            for k in (
                "kind",
                "label",
                "male_cns",
                "biological_learning",
                "expert_in_loop",
                "aileron",
                "elevator",
                "rudder",
                "throttle",
                "gru_hidden_norm",
                "gru_hidden_mean",
                "decoded_neural_step",
                "slew_alpha",
            )
            if k in snapshot.debug
        }
    if snapshot.observing or snapshot.ui_mode == "fly_control":
        payload["fly_observing"] = snapshot.fly_observing
    return payload

File: frontend/src/protocol.ts
export type EpisodeStatus =
  | "in_progress"
  | "landed"
  | "crashed"
  | "out_of_bounds"
  | "failed_approach";

export type ControllerName = "manual" | "expert" | "expert_observing";

export interface RunwayInfo {
  name: string;
  lat_deg: number;
  lon_deg: number;
  alt_m: number;
  heading_deg: number;
  length_m: number;
  width_m: number;
}

export interface HelloMessage {
  type: "hello";
  milestone: number;
  controller: ControllerName | string;
  control_authority?: string;
  physics: string;
  aircraft: string;
  runway: RunwayInfo;
  approach: {
    distance_m: number;
    agl_m: number;
    airspeed_kts: number;
    flight_path_deg: number;
  };
  integrity: {
    authoritative_physics: string;
    visuals: string;
    male_cns: boolean;
    male_cns_observing?: boolean;
    fly_controls_aircraft?: boolean;
    expert_is_biological?: boolean;
    note: string;
  };
  schedule?: {
    physics_hz: number;
    vision_hz: number;
    neural_hz: number;
    state_broadcast_hz: number;
  };
}

export interface ExpertTelemetry {
  kind: string;
  label?: string;
  male_cns?: boolean;
  phase: string;
  target_airspeed_kts: number;
  glideslope_error_deg: number;
  glideslope_error_m: number;
  centerline_error_m: number;
  heading_error_deg?: number;
}

export interface FlyObservingTelemetry {
  controlling: boolean;
  label: string;
  mode?: string;
  neural_step?: number | null;
  sim_time_s?: number | null;
  n_spikes: number;
  spikes_per_sec: number;
  n_outgoing_edges?: number;
  spike_checksum?: string;
  n_r1r6: number;
  n_descending: number;
  retina: {
    mean_luminance: number;
    mean_current: number;
    mean_temporal: number;
    left_mean_luminance: number;
    right_mean_luminance: number;
    current_sha256?: string;
  };
  descending: {
    n_descending: number;
    n_left: number;
    n_right: number;
    mean_hz: number;
    left_mean_hz: number;
    right_mean_hz: number;
    max_hz: number;
    window_steps: number;
  };
  visual_rates_hz?: Record<string, number>;
}

export interface StateMessage {
  type: "state";
  sim_time: number;
  paused: boolean;
  controller: ControllerName | string;
  control_authority?: string;
  observing?: boolean;
  spawn_seed?: number | null;
  position: {
    lat_deg: number;
    lon_deg: number;
    alt_msl_m: number;
    alt_agl_m: number;
    east_m: number;
    north_m: number;
    up_m: number;
    along_m: number;
    right_m: number;
  };
  velocity: {
    airspeed_kts: number;
    groundspeed_kts: number;
    vertical_speed_fpm: number;
  };
  attitude: {
    pitch_deg: number;
    roll_deg: number;
    heading_deg: number;
    alpha_deg: number;
    p_deg_s?: number;
    q_deg_s?: number;
    r_deg_s?: number;
    beta_deg?: number;
  };
  controls: {
    aileron: number;
    elevator: number;
    rudder: number;
    throttle: number;
  };
  surfaces?: {
    elevator_pos: number;
    aileron_pos: number;
    rudder_pos: number;
    throttle_pos: number;
  };
  gear: {
    on_ground: boolean;
    wow: boolean[];
  };
  episode: {
    status: EpisodeStatus;
    reason: string | null;
    touchdown_fpm: number | null;
  };
  expert?: ExpertTelemetry;
  fly_observing?: FlyObservingTelemetry;
}

export type ServerMessage = HelloMessage | StateMessage;

export interface PilotControls {
  aileron: number;
  elevator: number;
  rudder: number;
  throttle: number;
}
